extensions/cloud/portability-cloud-google/src/main/java/org/datatransferproject/cloud/google/GoogleCloudIdempotentImportExecutor.java

```diff
@@ -0,0 +1,233 @@
+  public <T extends Serializable> T executeOrThrowException(
+      String idempotentId, String itemName, Callable<T> callable) throws Exception {
+    Preconditions.checkNotNull(jobId, "executing a callable before initialization of a job");
+
+    if (knownValues.containsKey(idempotentId)) {
+      monitor.debug(
+          () ->
+              jobIdPrefix
+                  + format("Using cached key %s from cache for %s", idempotentId, itemName));
+      return (T) knownValues.get(idempotentId);
+    }
+
+    try {
+      T result = callable.call();
+      monitor.debug(
+          () -> jobIdPrefix + format("Storing key %s in cache for %s", idempotentId, itemName));
+      addResult(idempotentId, result);
+      return result;
+    } catch (Exception e) {
+      ErrorDetail errorDetail =
+          ErrorDetail.builder()
+              .setId(idempotentId)
+              .setTitle(itemName)
+              .setException(Throwables.getStackTraceAsString(e))
+              .build();
+      addError(idempotentId, errorDetail);
+      monitor.severe(() -> jobIdPrefix + "Problem with importing item: " + errorDetail);
+      throw e;
+    }
+  }
+
+  private <T extends Serializable> void addResult(String idempotentId, T result)
+      throws IOException {
+    knownValues.put(idempotentId, result);
+
+    try {
+      Transaction transaction = datastore.newTransaction();
+
+      transaction.put(createResultEntity(idempotentId, result));
+      if (errors.containsKey(idempotentId)) {
+        // if the errors contain this key, that means the ID
+        transaction.delete(getErrorKey(idempotentId));
+      }
+      transaction.commit();
+    } catch (DatastoreException e) {
+      monitor.severe(() -> jobIdPrefix + "Error writing result to datastore: " + e);
+    }
+  }
+
+  private void addError(String idempotentId, ErrorDetail errorDetail) throws IOException {
+    errors.put(idempotentId, errorDetail);
+    try {
+      Transaction transaction = datastore.newTransaction();
+      transaction.put(createErrorEntity(idempotentId, errorDetail));
+      transaction.commit();
+    } catch (DatastoreException e) {
+      monitor.severe(() -> jobIdPrefix + "Error writing ErrorDetails to datastore: " + e);
+    }
+  }
+
+  @Override
+  public <T extends Serializable> T getCachedValue(String idempotentId)
+      throws IllegalArgumentException {
+    if (!knownValues.containsKey(idempotentId)) {
+      throw new IllegalArgumentException(
+          idempotentId
+              + " is not a known key, known keys: "
+              + Joiner.on(", ").join(knownValues.keySet()));
+    }
+    return (T) knownValues.get(idempotentId);
+  }
+
+  @Override
+  public boolean isKeyCached(String idempotentId) {
+    return knownValues.containsKey(idempotentId);
+  }
+
+  @Override
+  public Collection<ErrorDetail> getErrors() {
+    return ImmutableList.copyOf(errors.values());
+  }
+
+  @Override
+  public void setJobId(UUID jobId) {
+    this.jobId = jobId;
```

Comment:
Should we check not null since these are lazy intialized member vars

Reply:
Done